impl/restart_integration_test.go

```diff
@@ -0,0 +1,291 @@
+				require.NoError(rh.t, rh.dt2.Start(rh.testCtx))
+				rh.dt2.SubscribeToEvents(subscriber)
+				require.NoError(rh.t, rh.dt2.RestartDataTransferChannel(rh.testCtx, chId))
+			},
+		},
+	}
+
+	for name, tc := range tcs {
+		t.Run(name, func(t *testing.T) {
+			// CREATE HARNESS
+			rh := newRestartHarness(t)
+			defer rh.cancel()
+
+			// START DATA TRANSFER INSTANCES
+			rh.sv.ExpectSuccessPush()
+			require.NoError(t, rh.dt1.Start(rh.testCtx))
+			require.NoError(t, rh.dt2.Start(rh.testCtx))
+
+			// SETUP DATA TRANSFER SUBSCRIBERS AND SUBSCRIBE
+			finished := make(chan peer.ID, 2)
+			errChan := make(chan *peerError, 2)
+			sent := make(chan uint64, totalIncrements)
+			received := make(chan uint64, totalIncrements)
+			receivedTillNow := atomic.NewInt32(0)
+
+			// counters we will check at the end for correctness
+			opens := atomic.NewInt32(0)
+			var finishedPeersLk sync.Mutex
+			var finishedPeers []peer.ID
+			disConnChan := make(chan struct{})
+
+			var subscriber datatransfer.Subscriber = func(event datatransfer.Event, channelState datatransfer.ChannelState) {
+				if event.Code == datatransfer.DataSent {
+					if channelState.Sent() > 0 {
+						sent <- channelState.Sent()
+					}
+				}
+
+				// disconnect and unlink the peers after we've received the required number of increments
+				if event.Code == datatransfer.DataReceived {
+					if channelState.Received() > 0 {
+						receivedTillNow.Inc()
+						received <- channelState.Received()
+						if receivedTillNow.Load() == int32(tc.stopAt) {
+							require.NoError(t, rh.gsData.Mn.UnlinkPeers(rh.peer1, rh.peer2))
+							require.NoError(t, rh.gsData.Mn.DisconnectPeers(rh.peer1, rh.peer2))
+							disConnChan <- struct{}{}
+						}
+					}
+				}
+
+				if channelState.Status() == datatransfer.Completed {
+					finishedPeersLk.Lock()
+					finishedPeers = append(finishedPeers, channelState.ManagerPeer())
+					finishedPeersLk.Unlock()
+					finished <- channelState.ManagerPeer()
+				}
+				if event.Code == datatransfer.Error {
+					err := xerrors.New(channelState.Message())
+					errChan <- &peerError{channelState.ManagerPeer(), err}
+				}
+				if event.Code == datatransfer.Open {
+					opens.Inc()
+				}
+
+				if event.Code == datatransfer.Restart {
+					fmt.Printf("\n Restart for peer %s", channelState.ManagerPeer().Pretty())
+				}
+			}
+			rh.dt1.SubscribeToEvents(subscriber)
+			rh.dt2.SubscribeToEvents(subscriber)
+
+			// OPEN PUSH
+			chid := tc.openPushF(rh)
+			// wait for disconnection to happen
+			<-disConnChan
+			t.Logf("peers unlinked and disconnected, total increments received till now: %d", receivedTillNow.Load())
+
+			// Define function to wait for data transfer to complete
+			waitF := func(wait time.Duration, nCompletes int) (sentI, receivedI []uint64, err error) {
+				completes := 0
+
+				waitCtx, cancel := context.WithTimeout(rh.testCtx, wait)
+				defer cancel()
+				i := 0
```

Comment:
tad silly but is this variable actually used?

Reply:
Great catch, not used anywhere. Have removed it.